enterprise/backend/test/metabase_enterprise/semantic_search/indexer_test.clj

```diff
@@ -0,0 +1,454 @@
+              indexed        (atom [])]
+          (with-redefs [semantic.indexer/sleep       (fn [_])
+                        semantic.index/upsert-index! (fn [pgvector index docs]
+                                                       (let [docs (vec docs)]
+                                                         (swap! indexed into docs)
+                                                         (upsert-index! pgvector index docs)))]
+            ;; need some data first for early exit, or it does nothing
+            (semantic.gate/gate-documents! pgvector index-metadata [(version c1 t1)])
+            (with-open [loop-thread (open-loop-thread! pgvector
+                                                       index-metadata
+                                                       index
+                                                       indexing-state)]
+              (vswap! indexing-state assoc :exit-early-cold-duration (Duration/ofMillis 1))
+              (let [{:keys [^Thread thread caught-ex]} @loop-thread
+                    max-wait (+ (System/currentTimeMillis) 100)]
+                (testing "exits"
+                  (is
+                   (loop []
+                     (cond
+                       (not (.isAlive thread)) true
+                       (< max-wait (System/currentTimeMillis)) false
+                       :else (recur))))
+                  (testing "has seen our row before exiting"
+                    (is (= 1 (count @indexed))))
+                  (testing "did not wait too long"
+                    (is (<= (- (System/currentTimeMillis) run-time) 50)))
+                  (testing "did not crash"
+                    (is (nil? @caught-ex))))))))))))
+
+(deftest init-indexing-state-test
+  (testing "initializes indexing state from metadata row"
+    (let [metadata-row {:indexer_last_poll (Instant/parse "2025-01-01T12:00:00Z")
+                        :indexer_last_seen (Instant/parse "2025-01-01T11:30:00Z")}
+          state        (semantic.indexer/init-indexing-state metadata-row)
+          state-value  @state]
+
+      (is (= (Instant/parse "2025-01-01T12:00:00Z") (get-in state-value [:watermark :last-poll])))
+      (is (= (Instant/parse "2025-01-01T11:30:00Z") (get-in state-value [:watermark :last-seen])))
+      (is (map? (:candidate-filter state-value)))
+      (is (zero? (:last-indexed-count state-value)))
+      (is (zero? (:last-poll-count state-value))))))
+
+(deftest quartz-job-run!-test
+  (let [pgvector        semantic.tu/db
+        index-metadata  (semantic.tu/unique-index-metadata)
+        open-job-thread (fn [& args]
+                          (let [caught-ex (volatile! nil)]
+                            (semantic.tu/closeable
+                             {:caught-ex caught-ex
+                              :thread
+                              (doto (Thread.
+                                     (fn []
+                                       (loop []
```

Comment:
ditto I think this loop is redundant, clearly no `recur`. Double check below asserts hold regardless and if not why the hell did they not fail!

Reply:
2599b0586dd0